src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs

```diff
@@ -1,494 +1,380 @@
+            StructHasFieldOfType<PropertyDeclarationSyntax>();
+
+            var fields = @struct.DescendantNodes().OfType<FieldDeclarationSyntax>();
+            foreach (var field in fields) 
+            {
+                CheckFieldValidity(field, @struct.Identifier, userCreatedTypes);
+            }
+            // Public types must have some member
+            if (!fields.Any())
             {
-                Report(ref context, DiagnosticRules.StructHasInvalidFieldRule2, structDeclaration.GetLocation(), structDeclaration.Identifier,  SimplifySyntaxTypeString(typeof(T).Name));
-                return true;
+                Report(WinRTRules.StructWithNoFieldsRule, @struct.GetLocation(), @struct.Identifier);
             }
-            return false;
         }
 
-        /// <summary>
-        /// returns true if there is a field declared private, 
-        /// or (inclusive) declared with a type that is a class or one of object, byte or dynamic</summary> 
-        /// <param name="context"></param><param name="field">The field to inspect</param><param name="structId">The name of the struct the field belongs to</param>
+        /// <summary>Struct fields must be public, cannot be const, or `object`, `dynamic`, or any defined class/interface type</summary> 
+        /// <param name="field">The field to inspect</param><param name="structId">The name of the struct the field belongs to</param>
         /// <param name="typeNames">A list of qualified class and interface names, which are invalid types to use in a struct for WinRT Component</param>
-        /// <returns>True if the struct has a field of an type that is not supported in Windows Runtime</returns>
-        public bool CheckFieldValidity(ref GeneratorExecutionContext context, 
+        private void CheckFieldValidity(
             FieldDeclarationSyntax field, 
             SyntaxToken structId, 
             HashSet<INamedTypeSymbol> typeNames)
         {
-            bool found = false; 
 
-            /* No private fields allowed in Windows Runtime components */
             if (!IsPublic(field))
             { 
-                Report(ref context, DiagnosticRules.StructHasPrivateFieldRule, field.GetLocation(), structId);
-                found |= true;
+                Report(WinRTRules.StructHasPrivateFieldRule, field.GetLocation(), structId);
             }
 
-            if (ModifiersContains(field.Modifiers, "const"))
+            if (field.Modifiers.Where(m => m.IsKind(SyntaxKind.ConstKeyword)).Any())
             {
-                Report(ref context, DiagnosticRules.StructHasConstFieldRule, field.GetLocation(), structId);
-                found |= true;
+                Report(WinRTRules.StructHasConstFieldRule, field.GetLocation(), structId);
             }
 
+            // helper function to see if a string is in the list of typenames
+            bool SymbolSetHasString(string typeStr) { return typeNames.Where(sym => sym.ToString().Contains(typeStr)).Any(); }

```

Comment:
To confirm, is this looking for a partial match due to contains?

Reply:
I found an alternative approach that doesn't use this kind of check